File: Machine-Learning-A-Z/Part 4 - Clustering/Section 24 - K-Means Clustering/kmeans.py
# -*- coding: utf-8 -*-
"""
Created on Wed May 24 20:59:15 2017

@author: Arshad
"""
#Importing libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

#Importing Mall dataset with panda
dataset = pd.read_csv('Mall_Customers.csv')
X = dataset.iloc[:,[3,4]].values
                
#Using elbow mwthod to determine number of cluster
from sklearn.cluster import KMeans
wcss = []
for i in range(1,11):
    kmeans = KMeans(n_clusters=i, init='k-means++',max_iter = 10, n_init=10, random_state = 0)
    kmeans.fit(X)
    wcss.append(kmeans.inertia_)
plt.plot(range(1,11), wcss)
plt.title('The Elbow method')
plt.xlabel('Number of cluster')
plt.ylabel('WCSS')
plt.show()

#Applying Kmeans to null dataset
kmeans = KMeans(n_clusters=5, init='k-means++',max_iter = 300, n_init=10, random_state = 0)  
y_kmeans = kmeans.fit_predict(X)

plt.scatter(X[y_kmeans == 0, 0], X[y_kmeans == 0, 1], s=100, c='red',label='Cluster1')
plt.scatter(X[y_kmeans == 1, 0], X[y_kmeans == 1, 1], s=100, c='blue',label='Cluster1')
plt.scatter(X[y_kmeans == 2, 0], X[y_kmeans == 2, 1], s=100, c='green',label='Cluster1')
plt.scatter(X[y_kmeans == 3, 0], X[y_kmeans == 3, 1], s=100, c='cyan',label='Cluster1')
plt.scatter(X[y_kmeans == 4, 0], X[y_kmeans == 4, 1], s=100, c='magenta',label='Cluster1')

plt.legend
plt.show()

File: Machine-Learning-A-Z/Part 4 - Clustering/Section 24 - K-Means Clustering/kmeans.R
#Kmeans Clustering

dataset <- read.csv('Mall_customers.csv')
X <- dataset[4:5]

#using the elbow method to fins nu number of clustering
set.seed(6)
wcss <- vector()
for (i in 1:10) wcss[i] <- sum(kmeans(X,i)$withinss)
plot(1:10, wcss, type = 'b', main = paste('Cluster of client'), xlab = 'Cluster of clients', ylab = 'wcss')

#Applying k-means to mall dataset
set.seed(29)
kmeans <- kmeans(X, 5, iter.max = 100,nstart = 10)

#Visulalizing the clusters
library(cluster)
clusplot(X,
         kmeans$cluster,
         lines = 0,
         shade = TRUE,
         color = TRUE,
         labels = 2,
         plotchar = FALSE,
         span = TRUE,
         main = paste('Cluster of clients'),
         xlab = 'Annual Income',
         ylab = 'Spending Score')